autobahn/wamp_tcp_client.hpp

```
@@ -0,0 +1,111 @@
+    public:
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const boost::asio::ip::tcp::endpoint &Endpoint,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(Endpoint),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        wamp_tcp_client(std::shared_ptr<boost::asio::io_service> pIo,
+                        const std::string &IpAddress,
+                        uint16_t Port,
+                        const std::string &Realm,
+                        bool Debug = false) :
+                m_realm(Realm),
+                m_rawsocket_endpoint(boost::asio::ip::address::from_string(IpAddress), Port),
+                m_pIo(pIo),
+                m_debug(Debug) {
+            init();
+        }
+
+        /**
+     * @brief initialize members, creating shared pointers for the socket, and session
+     */
+        void init() {
+            m_pSession.reset();
+            m_pSocket.reset();
+            m_connected = boost::promise<bool>();
+            m_pSocket = std::make_shared<boost::asio::ip::tcp::socket>(*m_pIo);
+            m_pSession = std::make_shared<wamp_tcp_session_t>(*m_pIo, *m_pSocket, *m_pSocket, m_debug);
+        }
+
+        /**
+     * @brief launches the session asynchronously, returns a future containing an error code (0 means success)
+     */
+        boost::future<bool> launch() {
+            m_pSocket->async_connect(m_rawsocket_endpoint, [&](boost::system::error_code ec) {
+                if (!ec) {
+                    m_startFuture = m_pSession->start().then([&](boost::future<bool> started) {
+                        m_joinFuture = m_pSession->join(m_realm).then([&](boost::future<uint64_t> connected) {
+                            m_connected.set_value(true);
```

Should the session ID be returned instead of true/false? I mean, I'm sure `wamp_session` has a reason why it returns a uint64_t instead of a bool, maybe this behaviour should be preserved here. I guess that also means making the lambda throw instead of setting `m_connected` to false, which is the standard way of making exceptions abort instead of continuing with the next continuation.